zbus/src/message/header.rs

```diff
@@ -314,6 +309,15 @@ impl<'m> Header<'m> {
 
 static SERIAL_NUM: AtomicU32 = AtomicU32::new(0);
 
+/// Get a new guaranteed-unique serial.
+pub fn take_serial() -> NonZeroU32 {
```

Comment:
I don't see a reason to expose this. If the user wants control of the serial numbers, then they are very much on their own for it.

Reply:
In order to guarantee non-collision both zbus and the user need to use serials from the same pool.

Reply:
> In order to guarantee non-collision both zbus and the user need to use serials from the same pool.

So you want to generate your own serials on the same connection while zbus also generates them for you? Why? :thinking:

Reply:
I mean, why not always generate them for all the messages for the connection in question? I'm not sure I want to support this. It will still be on the user to ensure that the serial numbers are unique. I'd rather they either let zbus handle all serial numbers or they do it, not both.

Reply:
Zbus always generates messages... during connection setup, etc.  Is there a reason you don't want to expose this?

Reply:
> Zbus always generates messages... during connection setup, etc.

There is only 1 message during the handshake and you can safely assume that to be `1` (or you can start from say `50` to be sure there is no conflict with that) but even if you had a conflict with the handshake `Hello` message, it's not a problem (or at least it should not be) since that method has already been replied to and the spec doesn't disallow re-use of serial numbers.

> Is there a reason you don't want to expose this?

It's a good practice to keep the API surface small for multiple reasons so new API should ideally be justified with compelling use cases, more so if it's low-level API most people would never need.

Reply:
> for multiple reasons

Which reasons?

Reply:
Done.

Reply:
> Which reasons? 

More maintainance burdon (the public API is then a contract between the maintainer and the user and can't be broken w/o major version bumps), reduced security because of larger attack surface, more API resulting in users having a harder time finding the API they need (i-e worse user experience) etc etc.

> Done

Thank you.

Reply:
> but are all implementations fine with it?

The only implementation that had a problem with it, was xdg-dbus-proxy and that was a considered a bug in the proxy and fixed several months ago. Although I'm not sure if the reuse would have been problematic for the buggy versions even, they had problem with serial numbers not being monotonically increasing.

The important bit is that there is nothing in the spec that would even imply that the serial numbers can't be reused. How would it even work? It's a 32-bit number and will have to be wrapped around at some point.